File: JavaModule/app/src/main/java/java/module/labyrintos/Permission/PermissionActivity.java
package java.module.labyrintos.Permission;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import java.module.labyrintos.R;

public class PermissionActivity  extends AppCompatActivity {
    private final int REQUEST_PERMISSION = 1999;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        int permssionCheck = ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION);
        //권한 승인이 되어있지 않을경우
        //if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
        if (permssionCheck != PackageManager.PERMISSION_GRANTED) {
            Toast.makeText(this, "권한 승인이 필요합니다", Toast.LENGTH_LONG).show();
            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
                    REQUEST_PERMISSION);
            if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.ACCESS_FINE_LOCATION)) {
                Toast.makeText(this,"이전에 거절한적이 있음",Toast.LENGTH_LONG).show();
            } else {
                Toast.makeText(this,"신규",Toast.LENGTH_LONG).show();
            }
        }else{
            //do
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        switch (requestCode) {
            case REQUEST_PERMISSION: {
                // If request is cancelled, the result arrays are empty.
                if (grantResults.length > 0
                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    Toast.makeText(this, "권한이 허가됨.", Toast.LENGTH_LONG).show();
                    //do
                } else {
                    Toast.makeText(this, "권한 취소", Toast.LENGTH_LONG).show();
                    finish();
                }
                return;
            }

        }
    }
}

File: KotlinModule/app/src/main/java/kotlins/module/labyrintos/Permission/PermissionActivity.kt
package kotlins.module.labyrintos.Permission

import android.Manifest.permission.*
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import kotlins.module.labyrintos.R

/**
 * Created by Labyrintos on 2019-11-03
 */
abstract class PermissionActivity : AppCompatActivity() {
    //퍼미션 응답 처리 코드
    private val REQUEST_PERMISSION = 100

    abstract val requiredPermission: Array<String>

    abstract fun doSomething()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        if(checkPermissions()) doSomething()


    }
     fun checkPermissions() :Boolean{
        //API 23 이하일 경우 권한체크 필요없음
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) return true

        //거절당한 권한 목록 가져오기
        val rejectedPermission = requiredPermission.filter {
            ContextCompat.checkSelfPermission(this, it) == PackageManager.PERMISSION_DENIED
        }

        if (rejectedPermission.isNotEmpty()) {
            ActivityCompat.requestPermissions(
                this,
                rejectedPermission.toTypedArray(),
                REQUEST_PERMISSION
            )
            return false
        }
         return true
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (REQUEST_PERMISSION) {
            REQUEST_PERMISSION -> if (grantResults.isNotEmpty()) {
                if(grantResults.filter { it == PackageManager.PERMISSION_DENIED }.isNotEmpty()){
                   showAlertDialog("정상적으로 이용하려면 권한이 필요합니다.")
                }else{
                    doSomething()
                }
            }
        }
    }

    fun showAlertDialog(msg: String) {
        //권한 거부
        AlertDialog.Builder(this).apply {
            setMessage(msg)
            setPositiveButton("설정"){
                _,_->checkPermissions()
            }
        }.show()
    }

}